File: dash_docs/chapters/dash_datatable/virtualization/examples/virtualization.py
import dash
import dash_table
import pandas as pd


app = dash.Dash(__name__)

df = pd.read_csv('https://raw.githubusercontent.com/plotly/datasets/master/Emissions%20Data.csv').reset_index()
df['Emission'] = df['Emission'].map(lambda x: '{0:.2f}'.format(x))

app.layout = dash_table.DataTable(
        id='table-virtualization',
        data=df.to_dict('records'),
        columns=[
            {'name': i, 'id': i} for i in df.columns
        ],
        fixed_rows={ 'headers': True, 'data': 0 },
        style_cell={
            'whiteSpace': 'normal'
        },
        style_data_conditional=[
            {'if': {'column_id': 'index'},
             'width': '50px'},
            {'if': {'column_id': 'Year'},
             'width': '50px'},
            {'if': {'column_id': 'Country'},
             'width': '100px'},
            {'if': {'column_id': 'Continent'},
             'width': '70px'},
            {'if': {'column_id': 'Emission'},
             'width': '75px'},
        ],
        virtualization=True,
        page_action='none'
)


if __name__ == '__main__':
    app.run_server(debug=True)

File: dash_docs/chapters/dash_datatable/part8/examples/virtualization.R
library(dash)
library(dashHtmlComponents)
library(dashTable)

app <- Dash$new()

df <- read.csv(
  file = 'https://raw.githubusercontent.com/plotly/datasets/master/Emissions%20Data.csv',
  stringsAsFactors = TRUE
)

df <- cbind(index = 1:dim(df)[1], df)

app$layout(
  dashDataTable(
    id = 'table-virtualization',
    data = df_to_list(df),
    columns = lapply(colnames(df),
                     function(column) {
                       list(name = column,
                            id = column)
                     }),
    editable = TRUE,
    fixed_rows = list(headers = TRUE, data = 0),
    style_cell = list(whiteSpace = 'normal'),
    style_data_conditional = list(
      list(
        'if' = list(column_id = 'index'),
        width = '50px'
      ),
      list(
        'if' = list(column_id = 'Year'),
        width = '50px'
      ),
      list(
        'if' = list(column_id = 'Country'),
        width = '100px'
      ),
      list(
        'if' = list(column_id = 'Continent'),
        width = '70px'
      ),
      list(
        'if' = list(column_id = 'Emission'),
        width = '75px'
      )
    ),
    virtualization = TRUE,
    page_action = 'none'
  )
)

app$run_server()
